DOC-14385-migrate-system-collection (#974)
couchbaselabs/docs-sync-gateway: modules/ROOT/pages/migrate-metadata-system-collection.adoc

* {sgw} metadata no longer appears alongside user-modifiable documents in the Capella UI.
* Eventing functions and {sqlpp} queries running against the default collection no longer process {sgw} internal documents.

[IMPORTANT]
====
This migration is opt-in and is never applied automatically at upgrade.
Your existing deployment is unaffected until you enable the feature.

The opt-in configuration flags are one-way.
Once enabled, the migration cannot be reversed.
There is no supported path to move metadata back to `_default._default`.
====

== Prerequisites

* {sgw} 4.1 or later on all nodes in the cluster.
* Couchbase Server version that supports the `_system._mobile` collection.
// ⚠️ Confirm minimum CBS version with Ben Brooks before publishing.
* The `_system._mobile` collection must be available on the target bucket.
{sgw} rejects the opt-in configuration if this collection is unavailable.

== How the Migration Works

The migration uses a two-level opt-in: one at the cluster level and one per database.

=== Cluster-level flag

The bootstrap configuration flag `use_system_metadata_collection` acts as the cluster-wide feature gate.

> Enabling it allows individual databases to opt in and enables the dual-collection read behaviour at bootstrap.
> By itself, enabling the cluster flag does not migrate any data.

Reviewer: The bootstrap-level option (added primarily for Capella App Services - but on-prem users _could_ decide to use this) auto-opts in all databases when they load. Not this behaviour as described.

```suggestion
Enabling it automatically opts in all databases to use the system metadata collection, and trigger migration.
```

include::ROOT:partial$_set_page_context.adoc[]

:param-abstract:
:topic-group: Application Deployment
:param-related: {release-notes--xref} | {compatibility--xref}
include::ROOT:partial$_show_page_header_block.adoc[]

== Overview

{sgw} stores internal metadata, including user documents, role documents, session documents, replication state, and DCP checkpoints, in the default collection (`_default._default`).
This couples {sgw}'s internal bookkeeping with customer document data in the same collection.

{sgw} 4.1 introduces an opt-in migration that relocates this metadata to the system collection (`_system._mobile`).
Migrating provides the following benefits:

* Metadata is isolated from user application data.
* {sgw} metadata no longer appears alongside user-modifiable documents in the Capella UI.
* Eventing functions and {sqlpp} queries running against the default collection no longer process {sgw} internal documents.

[IMPORTANT]
====
This migration is opt-in and is never applied automatically at upgrade.
Your existing deployment is unaffected until you enable the feature.

> 

Reviewer: As mentioned in the other PR - making a note of required cluster compatibility version alongside the feature is a good idea - Maybe here?

Will leave specific wording and possible linking to the cluster compatibility page to you.

```
The migration feature is only supported when the full Sync Gateway cluster has been upgraded to 4.1 and is running with a cluster compatibility version of 4.1
```

[source,json]
----
{
  "bootstrap": {
    "use_system_metadata_collection": true
  }
}
----

This is a one-way change.
Once set to `true`, the flag cannot be set back to `false`.

=== Step 2: Opt in each database

Add the following to the database configuration for each database you want to migrate:

[source,json]
----
{
  "use_system_metadata_collection": true
}
----

This is also a one-way change per database.

After you apply the database configuration, {sgw} starts the migration automatically once the configuration is confirmed on all nodes.
No further action is required to start the migration.

== Monitor Migration Progress

Use the `GET /{db}/_metadata_migration` endpoint on the Admin REST API to check the status of a migration.

NOTE: This endpoint is available on the Admin port (`:4985`) only.

*Example request:*

[source,bash]
----
GET /{db}/_metadata_migration
----

*Example response:*

[source,json]
----
{
  "status": "running",
  "migration_id": "a1b2c3d4-...",
  "docs_processed": 45200,
  "docs_attempted": 45200,
  "docs_failed": 0,
  "docs_out_of_scope": 12,
  "passes": 2
}
----

[cols="1,3"]
|===
| Field | Description

| `status`

> | Current migration status: `running`, `completed`, or `error`.

Reviewer: ```suggestion
| Current migration status: `running`, `completed`, `error`, `stopping`, `stopped`.
```

These are left in place.

| `passes`
| Number of scan passes the migrator has completed.
The migration is complete when a pass finds no remaining in-scope keys.
|===

== Manual Migration Controls

The migration starts automatically after opt-in.
Use the `POST /{db}/_metadata_migration` endpoint to intervene manually if needed.

*Required RBAC role:* Sync Gateway Architect

[cols="1,3"]
|===
| Query parameter | Description

| `action`
| `start` (default) to start or resume the migration, or `stop` to halt it.

| `reset`
| Set to `true` to discard previous progress and start a fresh migration run.
|===

*Retry after an error:*

[source,bash]
----
POST /{db}/_metadata_migration?action=start
----

*Stop a running migration:*

[source,bash]
----
POST /{db}/_metadata_migration?action=stop
----

[NOTE]
====
A `400` response is returned if the database has not opted in to `use_system_metadata_collection`.
====

For the full API reference, see xref:ROOT:rest_api_admin_static.adoc#tag/Database-Management/operation/get_db-_metadata_migration[GET /{db}/_metadata_migration].

== Prometheus Statistics

> {sgw} 4.1 adds seven Prometheus metrics under the `sgw_metadata_migration` subsystem, each labelled per database.

Reviewer: There was one more stat added towards the end of development

```suggestion
{sgw} 4.1 adds eight Prometheus metrics under the `sgw_metadata_migration` subsystem, each labelled per database.
```

[cols="1,1,3"]
|===
| Metric | Type | Description

| `sgw_metadata_migration_docs_scanned_total`
| Counter
| Total fallback keys observed across all scan passes.

| `sgw_metadata_migration_docs_migrated`
| Counter
| Documents successfully moved or removed from the fallback collection.

| `sgw_metadata_migration_docs_out_of_scope`
| Gauge
| Keys in the fallback collection belonging to other databases or bucket-level bootstrap documents on the most recent pass.
This is a snapshot of the last pass, not a cumulative total.

| `sgw_metadata_migration_docs_unknown_prefix`
| Gauge
| Unrecognized keys in the fallback collection on the most recent pass.
A persistently non-zero value indicates the migration will not complete without intervention.
Contact Couchbase Support if this value remains non-zero.

| `sgw_metadata_migration_errors`
| Counter
| Per-document errors during move or delete operations.

| `sgw_metadata_migration_seq_poison_pill_applied`
| Counter
| Number of times this node applied the sequence counter handoff during migration.
Typically `0` or `1` per migration run.

| `sgw_metadata_migration_passes`
| Counter
| Number of scan passes executed.

> |===

Reviewer: There was one more stat added towards the end of development:

```suggestion

| `sgw_metadata_migration_abandoned_runs`
| Counter
| Number of migration runs that the manager abandoned after several retry attempts.
|===
```
